apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/QueryTest.CreateDocumentSnapshotComparer.cs

```diff
@@ -0,0 +1,209 @@
+                CreateSnapshot("doc2", 8, 9),
+                CreateSnapshot("doc4", 15, 15)
+            );
+        }
+
+        [Fact]
+        public void OrderByValueDescending_DocumentIdForTieBreak()
+        {
+            var query = s_db.Collection("col").OrderByDescending("value1");
+            // Last ordering was descending, so doc2 comes before doc1.
+            AssertDocsInOrder(query,
+                CreateSnapshot("doc4", 15, 15),
+                CreateSnapshot("doc2", 8, 9),
+                CreateSnapshot("doc1", 8, 5),
+                CreateSnapshot("doc3", 3, 1)
+            );
+        }
+
+        [Fact]
+        public void OrderByTwoValues()
+        {
+            var query = s_db.Collection("col").OrderBy("value1").OrderBy("value2");
+            AssertDocsInOrder(query,
+                CreateSnapshot("doc4", 3, 5),
+                CreateSnapshot("doc1", 8, 5),
+                CreateSnapshot("doc3", 8, 5),
+                CreateSnapshot("doc2", 8, 9),
+                CreateSnapshot("doc5", 15, 15)
+            );
+        }
+
+        [Fact]
+        public void OrderByTwoValues_SecondIsDescending()
+        {
+            var query = s_db.Collection("col").OrderBy("value1").OrderByDescending("value2");
+            AssertDocsInOrder(query,
+                CreateSnapshot("doc4", 3, 5),
+                CreateSnapshot("doc2", 8, 9),
+                CreateSnapshot("doc3", 8, 5),
+                CreateSnapshot("doc1", 8, 5),
+                CreateSnapshot("doc5", 15, 15)
+            );
+        }
+
+        private static void AssertDocsInOrder(Query query, params DocumentSnapshot[] docs)
+        {
+            var rng = new Random();
+            var comparer = query.CreateDocumentSnapshotComparer();
+            // This is slightly simpler than getting all permutations. Randomness in tests isn't great, but
+            // we'll almost certainly hit every permutation anyway, given the size we have.
+            for (int i = 0; i < 50; i++)
+            {
+                var unordered = Shuffle(docs, rng).ToList();
+                unordered.Sort(comparer);
+                // Assert based on ID rather than content, for cleaner error reporting.
+                Assert.Equal(docs.Select(doc => doc.Reference.Id), unordered.Select(doc => doc.Reference.Id));
+            }
+        }
+
+        public static IEnumerable<T> Shuffle<T>(IEnumerable<T> source, Random rng)
+        {
+            T[] elements = source.ToArray();
+            // Note i > 0 to avoid final pointless iteration
+            for (int i = elements.Length - 1; i > 0; i--)
+            {
+                // Swap element "i" with a random earlier element it (or itself)
+                int swapIndex = rng.Next(i + 1);
+                T tmp = elements[i];
+                elements[i] = elements[swapIndex];
+                elements[swapIndex] = tmp;
+            }
+            // Lazily yield (avoiding aliasing issues etc)
```

Comment:
I don't see why you can't just return `elements`? Nothing else has a reference to `elements`?

Reply:
Oh yes, I can in this case. This was copied from one of my Stack Overflow answers :)